### PR TITLE
Plan,TextMemorySkill improvements and custom planner examples

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
### Motivation and Context
The motivation of the PR is to demonstrate custom planning capabilities, specifically for the markup returned by `ContextQuery` function. 

In creating this example, some gaps were found in the Plan object and TextMemorySkill. This PR proposes some changes to those objects.

#### Plan
- Plan now updates its state with the result of each step, appending the result to the previous result if the step has more than one named output. This allows the plan to keep track of the context and the intermediate results of each step, and to pass them to the next step as input.

#### Example Custom|Markup Planner
- A new example for using a custom planner to create and execute a plan using the Markup Skill, which can process XML markup created by the ContextQuery skill, which can generate natural language queries from a context. 
- A new example of using a custom planner to generate plans from XML markup, using a simple markup language that supports lookup, echo, and concat functions, as well as nested elements.
- Some minor fixes and improvements to the plan execution logic, the plan parser, the plan invocation logic, and the PlanToString method.

#### TextMemorySkill improvements (for the example).
- TextMemorySkill now has a constructor that allows customizing the default parameters for the Recall function, such as the collection, relevance, and limit. By default, getting a single memory for ContextQuery was not interesting. This change makes it easier to demonstrate more interesting capabilities.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:
- [x] Integration Tests pass https://github.com/lemillermicrosoft/semantic-kernel/actions/runs/4831594273

### CODE REVIEW

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs [hunk 1]
```diff
@@ -46,6 +46,19 @@ public class TextMemorySkill
     private const string DefaultRelevance = "0.75";
```
Comment (awharrison-28) [inline dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:46]: I like the constructor change here. I think with this, we should make DefaultRelevance = "0.0", 0.75 is very ADA specific

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs [hunk 2]
```diff
@@ -92,21 +105,22 @@ public async Task<string> RetrieveAsync(SKContext context)
...
-        var limit = context.Variables.ContainsKey(LimitParam) ? context[LimitParam] : DefaultLimit;
+        var limit = context.Variables.ContainsKey(LimitParam) ? context[LimitParam] : this._limit;
         if (string.IsNullOrWhiteSpace(limit)) { limit = DefaultLimit; }
 
         context.Log.LogTrace("Searching memories in collection '{0}', relevance '{1}'", collection, relevance);
 
         // TODO: support locales, e.g. "0.7" and "0,7" must both work
-        int limitInt = int.Parse(limit, CultureInfo.InvariantCulture);
+        var limitInt = int.Parse(limit, CultureInfo.InvariantCulture);
+        var relevanceScore = float.Parse(relevance, CultureInfo.InvariantCulture);
```
Comment (awharrison-28) [inline dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:121]: nit: call this relevanceThreshold? Or minimumRelevanc?

File: samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+    public static implicit operator XmlNode(XmlNodeInfo info)
+    {
+        return info.Node;
+    }
+}
+#pragma warning restore CA1815
+
+#pragma warning disable CA1711
+public static class XmlEx
```
Comment (dluc) [inline samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs:293]: why are we still using XML and not leveraging the plan data structure?

**Review by dluc — COMMENTED** [review]
after all the effort put into abstracting away from XML and having a plan data structure, we show an example with 200 lines of code dealing with XML. Does it mean that the plan data structure doesn't provide what we needed? Why can't we leverage all the work done instead of writing another parser?

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 4]
```diff
@@ -43,12 +43,19 @@ public sealed class Plan : ISKFunction
     public ContextVariables NamedParameters { get; set; } = new();
 
     /// <summary>
-    /// Named outputs for the function
+    /// Named outputs for the function -- these are used within a plan.
     /// </summary>
     [JsonPropertyName("named_outputs")]
     [JsonConverter(typeof(ContextVariablesConverter))]
     public ContextVariables NamedOutputs { get; set; } = new();
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:50]: I see a definite point of confusion between NamedOutputs and NamedResults. I think NamedOutputs could be changed, since it feels to me like "Plan.NamedOutputs" should be the outputs of the plan.

Maybe PlanVariables?

File: samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    private static string PlanToString(Plan originalPlan, string indent = " ")
+    {
+        var sb = new StringBuilder($"{indent}Goal: {originalPlan.Description}\n\n{indent}Steps:\n");
+
+        foreach (var step in originalPlan.Steps)
+        {
+            sb.Append(step.Steps.Count switch
+            {
+                0 =>
+                    $"{indent}{indent}- {string.Join(".", step.SkillName, step.Name)} {string.Join(" ", step.NamedParameters.Select(p => $"{p.Key}='{p.Value}'"))}{(step.NamedOutputs.Where(s => s.Key.ToUpper(System.Globalization.CultureInfo.CurrentCulture) != "INPUT").Select(p => $"{p.Key}").FirstOrDefault() is var namedOutputs ? $" => {namedOutputs}" : "")}",
```
Comment (shawncal) [inline samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs:173]: Can you please just break this out into a private helper function, or a mutli-line body? Break the magic up into variables and then fortmat the string, or use StringBuilder.

File: samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs (hunk 3 above)
Comment (shawncal) [reply samples/dotnet/kernel-syntax-examples/Example29_CustomPlanner.cs:293]: This literally is a "XmlMarkup" Plan example, showing how you might create a Plan (data structure) out of your own XML. I would think XML parsing would be a required component of such an example.

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 4 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:50]: Yeah @RogerBarreto and I had discussions on this last week that we need to continue. I wanted to push this change as a step in a better direction, but will pump the breaks for now.

The problem we want to solve is:
- How to identify that a plan/ISKFunction has an output that needs to be saved to the parent Plan.State
- How to identify that M<=N steps in a Plan have an output that is part of the parent output (Plan.State.ToString()).

Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:50]: NamedOutputs and NamedResults are both outputs of the plan. Right now, functions have a single output. In the future that could change where a function has multiple outputs that should be preserved in the state, the main issue is knowing which (single/multiple) output is something that should be part of the default result.

**Comment by lemillermicrosoft** [comment]
> after all the effort put into abstracting away from XML and having a plan data structure, we show an example with 200 lines of code dealing with XML. Does it mean that the plan data structure doesn't provide what we needed? Why can't we leverage all the work done instead of writing another parser?

Great idea that we should build a planner or example that uses the serialization syntax for plan creation.

As Shawn noted, the goal of this PR is to create a planner for existing semantic function `ContextQuery`.

File: samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs [hunk 6]
```diff
@@ -130,12 +131,21 @@ private static async Task BookSamplesAsync()
...
+        foreach (var step in originalPlan.Steps)
+        {
+            sb.Append(step.Steps.Count switch
+            {
+                0 =>
+                    $"{indent}{indent}- {string.Join(".", step.SkillName, step.Name)} {string.Join(" ", step.NamedParameters.Select(p => $"{p.Key}='{p.Value}'"))}{(step.NamedOutputs.Where(s => s.Key.ToUpper(System.Globalization.CultureInfo.CurrentCulture) != "INPUT").Select(p => $"{p.Key}").FirstOrDefault() is var namedOutputs ? $" => {namedOutputs}" : "")}",
+                _ => PlanToString(step, indent + indent)
+            });
+        }
+
```
Comment (lemillermicrosoft) [inline samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs:147]: note:

Comment (lemillermicrosoft) [reply samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs:147]: If I use both ActionPlanner and SequentialPlanner I must do the following. This is because ActionPlanner is a skill as well.

> var planner = new SequentialPlanner(kernel, new SequentialPlannerConfig() { ExcludedSkills = new[] { "this" } });

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs [hunk 7]
```diff
@@ -43,9 +43,22 @@ public class TextMemorySkill
     public const string LimitParam = "limit";
 
     private const string DefaultCollection = "generic";
-    private const string DefaultRelevance = "0.75";
+    private const string DefaultRelevance = "0.0";
```
Comment (dluc) [inline dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:46]: this seems a pretty big change, how so?

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs [hunk 8]
```diff
@@ -43,9 +43,22 @@ public class TextMemorySkill
...
     private const string DefaultLimit = "1";
 
+    /// <summary>
+    /// Creates a new instance of the TextMemorySkill
+    /// </summary>
+    /// <param name="collection">The default collection for Recall</param>
+    /// <param name="relevance">The default relevance value for Recall</param>
+    /// <param name="limit">The default limit for Recall</param>
+    public TextMemorySkill(string collection = DefaultCollection, string relevance = DefaultRelevance, string limit = DefaultLimit)
```
Comment (dluc) [inline dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:55]: users will need some help with the relevance param, e.g. what's the range? what's a good value? the function should probably validate it too

File: samples/skills/QASkill/ContextQuery/skprompt.txt [hunk 9]
```diff
@@ -17,27 +17,27 @@ essay: longer answers. You can have sub-elements such as fact and fiction
 TODAY is {{time.Date}}
 FIRST NAME: {{$firstname}}
 LAST NAME: {{$lastname}}
-CITYl {{$city}}
+CITY: {{$city}}
 STATE: {{$state}}
 COUNTRY: {{$country}}
-{{recall $input}}
+{{memory.recall $input}}
```
Comment (dluc) [inline samples/skills/QASkill/ContextQuery/skprompt.txt:23]: in all our demos we've always assumed there's a global "recall" function, which is a nice to have. Why this change?

File: samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs (hunk 6 above)
Comment (dluc) [reply samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs]: is there a way not to load the actionplanner skill in the kernel?

Comment (lemillermicrosoft) [reply samples/dotnet/kernel-syntax-examples/Example12_SequentialPlanner.cs]: Yes. But if you try to use both, this will happen.

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs (hunk 7 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:46]: Guidance was from @awharrison-28 in above comments. The current value is very much dependent on the ADA models, which may not be the default for users.

File: dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs (hunk 8 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/CoreSkills/TextMemorySkill.cs:55]: Added comments from below to help here. There hadn't been validation elsewhere -- if they give a value greater than 1.0 it will fail to find memories. I'll lean on existing flexibility there for now.

File: samples/skills/QASkill/ContextQuery/skprompt.txt (hunk 9 above)
Comment (lemillermicrosoft) [reply samples/skills/QASkill/ContextQuery/skprompt.txt:23]: I liked the consistency with `time` above but will revert.

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 10]
```diff
@@ -325,7 +346,7 @@ public FunctionView Describe()
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:346]: AddVariablesToContext(this.State, functionContext);
   - I *think* this can be moved above the loop. Looks like this has to only be called once before the first run, then the UpdateContextWithOutputs takes care of the remainder.

**Review by shawncal — APPROVED** [review]

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 10 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:346]: Not a blocker. Filed a task to check this out.

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 11]
```diff
@@ -252,18 +257,34 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
...
+                if (resultAppended)
                 {
                     continue;
                 }
 
-                this.State.Set(item.Key, result.Result.Trim());
+                this.State.Get(DefaultResultKey, out var currentPlanResult);
+                this.State.Set(DefaultResultKey, string.Join("\n", currentPlanResult.Trim(), resultValue));
+                resultAppended = true;
+                continue;
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:274]: `continue;` not needed

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 12]
```diff
@@ -252,18 +257,34 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
...
 
-            // Update state with named outputs (if any)
-            foreach (var item in step.NamedOutputs)
+            // Update Plan Result in State with matching outputs (if any)
+            bool resultAppended = false;
+            foreach (var item in this.Outputs.Intersect(step.Outputs))
             {
-                // ignore the input key
-                if (item.Key.ToUpperInvariant() == "INPUT")
+                if (resultAppended)
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:266]: once `resultAppended = true;` all the rest of the iterations will this this and `continue`, right?
Instead, just call `break;`?

But then it looks like resultAppended is always, unconditionally set to true at the end of the first iteration so maybe not a loop at all?

**Review by shawncal — CHANGES_REQUESTED** [review]
Pulled branch, tried it out. All works. I noticed a couple code things pointed out, worthy of another look.

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 10 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:338]: Yeah sounds good, will want to make sure I test this change well, for follow up.

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 13]
```diff
@@ -133,21 +131,27 @@ public Plan(ISKFunction function)
...
         IReadOnlyList<Plan> steps)
     {
         this.Name = name;
         this.SkillName = skillName;
         this.Description = description;
         this.NextStepIndex = nextStepIndex;
         this.State = state;
-        this.NamedParameters = namedParameters;
-        this.NamedOutputs = namedOutputs;
+        this.Parameters = parameters;
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:153]: `Outputs` is a much better name but I see you kept the `Parameters` instead of using `Inputs`.

My suggestion just to make it easier for the audience is where I see `Outputs` I would expect an `Inputs`.

`InputParameters` and `OutputParameters` would be my ask if you want to keep the `Parameters`

File: samples/dotnet/kernel-syntax-examples/RepoUtils/PlanExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        string goalHeader = $"{indent}Goal: {originalPlan.Description}\n\n{indent}Steps:\n";
+
+        string stepItems = string.Join("\n", originalPlan.Steps.Select(step =>
+        {
+            if (step.Steps.Count == 0)
+            {
+                string skillName = step.SkillName;
+                string stepName = step.Name;
+
+                string namedParams = string.Join(" ", step.Parameters.Select(param => $"{param.Key}='{param.Value}'"));
```
Comment (RogerBarreto) [inline samples/dotnet/kernel-syntax-examples/RepoUtils/PlanExtensions.cs:21]: update the `named` prefixes in those variables.

File: samples/dotnet/kernel-syntax-examples/Example31_CustomPlanner.cs [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.Planning;
+using Microsoft.SemanticKernel.SkillDefinition;
+using Microsoft.SemanticKernel.Skills.Web;
+using Microsoft.SemanticKernel.Skills.Web.Bing;
+using RepoUtils;
+
+// ReSharper disable CommentTypo
+// ReSharper disable once InconsistentNaming
+internal static class Example31_CustomPlanner
```
Comment (RogerBarreto) [inline samples/dotnet/kernel-syntax-examples/Example31_CustomPlanner.cs:20]: I think this example is awesome, but became too complex to fully understand.

I would split it into:
1. Creating a Custom Planner from a Custom Markup and Invoking it.
2. (Add new sample) Running Plans within Skills (Plan calling other Plans) 
3. Mixing the two above (Current example) with some reutilization.

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 16]
```diff
@@ -252,18 +257,26 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
             #region Update State
 
             // Update state with result
-            this.State.Update(result.Result.Trim());
+            this.State.Update(resultValue);
+
+            // Update Plan Result in State with matching outputs (if any)
+            if (this.Outputs.Intersect(step.Outputs).Any())
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:263]: I would suggest moving away of having a DefaultResultKey, since all the information is already in the plan.

Potentially I would add a `Results` property where you get the collection of results for the plan and you can actually do whatever you want with it, including Join.

File: dotnet/src/Extensions/Planning.SequentialPlanner/SequentialPlanParser.cs [hunk 17]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
 using System;
+using System.Collections.Generic;
 using System.Xml;
 using Microsoft.Extensions.Logging;
 using Microsoft.SemanticKernel.Orchestration;
```
Comment (RogerBarreto) [inline dotnet/src/Extensions/Planning.SequentialPlanner/SequentialPlanParser.cs:7]: In the `ToPlanFromXml` change the variable names `o`, `o2` to actually represent what they are.

**Review by RogerBarreto — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 18]
```diff
@@ -423,6 +436,32 @@ private static void AddVariablesToContext(ContextVariables vars, SKContext conte
...
+    /// <returns>The updated context.</returns>
+    private SKContext UpdateContextWithOutputs(SKContext context)
+    {
+        var resultString = this.State.Get(DefaultResultKey, out var result) ? result : this.State.ToString();
+        context.Variables.Update(resultString);
+
+        // copy previous step's variables to the next step
+        foreach (var item in this._steps[this.NextStepIndex - 1].Outputs)
+        {
+            if (this.State.Get(item, out var val))
```
[the remark this answers is not in the document]
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:452]: I'd like to not touch `ContextVariables` class as part of this change. The pattern is `TryGet` it's just named `Get`.

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 13 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:153]: Let's follow up separately on this one. I kept with `Parameters` since that's the naming on `SKFunction`

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 16 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/Plan.cs:263]: I'd really like to sticking with the State object and SKContext -- until those patterns change/evolve (SKRequest/SKResponse maybe?). Since Input/Result is overloaded in SKContext, having this special key is useful right now.

**Review by rogerio-barreto — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

File: samples/dotnet/kernel-syntax-examples/RepoUtils/PlanExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Linq;
+using Microsoft.SemanticKernel.Planning;
+
+namespace RepoUtils;
+
+internal static class PlanExtensions
+{
+    internal static string ToPlanString(this Plan originalPlan, string indent = " ")
```
Comment (shawncal) [inline samples/dotnet/kernel-syntax-examples/RepoUtils/PlanExtensions.cs:10]: Feature request (separate PR): please make sure Plan.ToString() calls this.
